instrumentation/instaredis/instaredis.go

```diff
@@ -0,0 +1,180 @@
+// (c) Copyright IBM Corp. 2022
+
+//go:build go1.13
+// +build go1.13
+
+package instaredis
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-redis/redis/v8"
+	instana "github.com/instana/go-sensor"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+)
+
+type commandCaptureHook struct {
+	options *redis.Options
+	sensor  *instana.Sensor
+}
+
+func setSpanCommands(span opentracing.Span, cmd redis.Cmder, cmds []redis.Cmder) {
+	if cmd != nil {
+		span.SetTag("redis.command", cmd.Name())
+
+		if cmd.Err() != nil {
+			span.SetTag("redis.error", cmd.Err().Error())
+			span.LogFields(otlog.Object("error", cmd.Err().Error()))
+		}
+	} else if len(cmds) > 0 {
+		var subCommands []string
+
+		if cmds[0].Name() == "multi" && cmds[len(cmds)-1].Name() == "exec" {
+			for i := 1; i < len(cmds)-1; i++ {
+				subCommands = append(subCommands, cmds[i].Name())
+			}
+		} else {
+			for _, cmd := range cmds {
+				subCommands = append(subCommands, cmd.Name())
+			}
+		}
+		span.SetTag("redis.command", "multi")
+		span.SetTag("redis.subCommands", subCommands)
+	}
+}
+
+func (h commandCaptureHook) handleHook(ctx context.Context, cmd redis.Cmder, cmds []redis.Cmder) {
+
+	connection := h.options.Addr
+
+	if connection == "FailoverClient" {
+		conn, err := h.options.Dialer(ctx, h.options.Network, "")
+
+		if err == nil {
+			connection = conn.RemoteAddr().String()
+		}
+	}
+
+	// if the IP provided in the Redis constructor have only ports. eg: :6379, :6380 and so on, we add 127.0.0.1
+	if strings.HasPrefix(connection, ":") {
+		connection = "127.0.0.1" + connection
+	}
```

Comment:
Also, can we figure out the address once while wrapping the client and not each time we execute a command?

Reply:
For some cases, it is possible to discover the connection earlier.
For some others, I depend on the context, which is available in a later stage, and provided by the hooks.
What I did in the last change was to try to recover the connection in cases where `ctx` is not necessary. 
For the other cases, the connection is cached once it is discovered. Commands still trigger the connection discovery, but as it will be cached after the first call to a command, the discovery will return much earlier after the first call to the hooks